src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java

```diff
@@ -154,29 +179,43 @@ public UUID getPendingRepair()
         return ref.get().getPendingRepair();
     }
 
-    public int getManifestSize()
-    {
-        return manifest.components().size();
-    }
-
     @Override
     public void write(StreamSession session, DataOutputStreamPlus out, int version) throws IOException
     {
+        boolean streamEntirely = shouldStreamEntireSSTable && out instanceof AsyncStreamingOutputPlus;
         SSTableReader sstable = ref.get();
-        CassandraStreamHeader.serializer.serialize(header, out, version);
-        out.flush();
 
-        if (shouldStreamEntireSSTable && out instanceof AsyncStreamingOutputPlus)
+        if (streamEntirely)
         {
-            CassandraEntireSSTableStreamWriter writer = new CassandraEntireSSTableStreamWriter(sstable, session, manifest);
-            writer.write((AsyncStreamingOutputPlus) out);
+            // Acquire lock to avoid concurrent sstable component mutation becaue of stats update or index summary
+            // redistribution, otherwise file sizes recorded in component manifest will be different from actual
+            // file sizes. (Note: Windows doesn't support atomic replace and index summary redistribution deletes
+            // existing file first)
+            // Recreate the latest manifest with hard links in case components are modified.
+            ComponentManifest manifest = sstable.runWithReadLock(ignored -> getComponentManifest(sstable, true));
```

Comment:
@jasonstack If we only hold the read lock for manifest creation, and we _don't_ create a hard link for the index summary component, what happens if, after we create the manifest, we start streaming, and then an index summary redistribution finishes and tries to save? It seems like creating a hard link for the index summary would work here, even if we don't do an atomic replace, given manifest creation would have to block if a summary saving is in progress anyway.

Reply:
so the index summary hark link is created on either old summary or new summary, never on the partially written file because of the lock..

Reply:
ah.. forgot to hard link `index summary`..